[Ho Pin Xian] iP

hopinxian opened the pull request.

Commit:
Add Gradle support

hopinxian committed:
Add greetings, echo, exit functionality

hopinxian committed:
Add add, list functionality

hopinxian committed:
Rename Duke.java as Butler.java

hopinxian committed:
Add Task.java

hopinxian committed:
Add complete task functionality

hopinxian committed:
Add DeadlineTask.java

hopinxian committed:
Add ToDoTask.java

hopinxian committed:
Add EventTask.java

hopinxian committed:
Add ToDos, Events, Deadlines functionality

hopinxian committed:
Add automated text UI testing

hopinxian committed:
Add exception handling functionality

hopinxian committed:
Add delete task functionality

hopinxian committed:
Add functionality to save task list on hard drive

hopinxian committed:
Add Butler.java updates (missed in previous commit) and taskList.txt

hopinxian committed:
Allow Butler to understand dates

hopinxian committed:
Merge branch 'branch-Level-7'

hopinxian committed:
Allow Butler to understand dates

hopinxian committed:
Add Ui.java, TaskList.java, Storage.java, Parser.java

Commit:
build.gradle: Update version to 8.29

raythx98 commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+        try {
+            switch (inputList[0]) {
+                case "todo":
+                    try {
```
Is the nesting too deep? Consider abstraction to your other classes

josiahkhoo replied:
I am with @raythx98 on this, perhaps you can make use of your Parser and Storage class?

hakujitsu replied:
I agree with @raythx98 and @josiahkhoo on this point as well, abstraction would decrease the LOC, making the method more readable.

In addition, could you abstract this method (and potentially other methods such as printTaskList) itself to the TaskList class?

hopinxian replied:
I have done the abstraction in my recent commits. Thank you for the comments.

raythx98 commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+                writeToFile(input);
+            }
+        } catch (Exception e) {
+            System.out.println("\nPlease provide a valid index.\n");
```
Can we assume that all thrown exceptions are only due to invalid index (and not other unchecked exceptions)?

hopinxian replied:
Thank you for the feedback. I have refactored my code such that exceptions caught are specific instead of blanket catching. This will allow me to be aware of other exceptions that may be thrown as well.

raythx98 commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    private static void completeTask(String input, ArrayList<Task> taskList, Boolean isNew) {
```
would a naming like "markTaskAsComplete" be clearer?

hopinxian replied:
Thank you for the feedback. I feel that completeTask would be a shorter name that also identifies the method purpose.

raythx98 commented on src/main/java/DeadlineTask.java:
```diff
@@ -0,0 +1,21 @@
+
+    public DeadlineTask(String summary, LocalDate deadline) {
+        super(summary);
+            this.deadline = deadline;
```
Unnecessary indentation

raythx98 commented on src/main/java/ToDoTask.java:
```diff
@@ -0,0 +1,11 @@
+public class ToDoTask extends Task {
```
Since ToDo is-a task, would a name like ToDo be sufficient?

hopinxian replied:
I feel that having the name of Task would better identify the class for other developers since they may not go deep to see the inheritance from Task class.

raythx98 reviewed (commented):
Generally, more abstraction is needed but code is generally understandable and follows the code style!

josiahkhoo commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+        String summary;
+        Boolean isValidCommand = false;
+        try {
+            switch (inputList[0]) {
+                case "todo":
+                    try {
+                        addedTask = new ToDoTask(inputList[1]);
+                        taskList.add(addedTask);
+                        reply += "\nI have added your ToDo task as follows:\n"
+                                + "Added: " + addedTask + "\n";
+                        isValidCommand = true;
+                    } catch (Exception e) {
+                        System.out.println("\nExcuse me, " +
+                                "but I will need a description for the ToDo task given.\n");
+                    }
+                    break;
+
```
Should there be indentations for case clauses?

hopinxian replied:
Thank you for pointing this out. I have removed the indentations to adhere to coding standard.

josiahkhoo commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+                        LocalDate formattedDeadline = LocalDate.parse(deadline);
+                        addedTask = new DeadlineTask(summary, formattedDeadline);
+                        taskList.add(addedTask);
+                        reply += "\nI have added your Deadline task as follows:\n"
+                                + "Added: " + addedTask + "\n";
+                        isValidCommand = true;
+                    } catch (DateTimeParseException e) {
+                        System.out.println("\nPlease input a valid Date format.\n");
+                    } catch (Exception e) {
+                        System.out.println("\nExcuse me, " +
+                                "please provide a summary and deadline using the tag /by.\n");
```
Perhaps you can standardize whether to break before an operator or after an operator?

hopinxian replied:
Yes I will.

josiahkhoo commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+        return summary;
+    }
+
+    public void markComplete() {
+        this.isComplete = true;
+    }
+
```
Should the setter method for isComplete follow the form of `void setComplete(boolean isComplete);`?

hopinxian replied:
I would consider this differently from a setter method since this method does not take in an argument. But I agree there would be more flexibility if this method is converted to a setter method.

josiahkhoo reviewed (commented):
I found the code generally clean and readable, perhaps abstracting some of the methods in Butler could help you with code organization?

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+        String input = sc.nextLine();
+
+        // Reply Loop
+        while(!input.equals("bye")) {
```
Should there be a space between while and the parathesis?

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    private static void writeToFile(String textToAdd) throws IOException {
```
Should this method be moved to the Storage class?

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+            } catch (IOException f) {
+                System.out.println("Code should never reach here." + f.getMessage());
+            }
+        }
```
Following the SLAP principle (Single Level of Abstraction Per method), perhaps this should be abstracted into a method of its own in the Storage class (e.g. populateTaskList)?

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+                    } catch (Exception e) {
+                        System.out.println("\nExcuse me, " +
+                                "please provide a summary and time of event using the tag /at.\n");
+                    }
```
I noticed multiple try/catch blocks embedded in this method. Would it be possible to catch all the different types of exceptions in the most external try/catch block and handle them respectively, to improve code organisation?

hopinxian replied:
Ah yes, that is a great idea to improve readability. Thank you.

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+        System.out.println(greetings);
+
+        ArrayList<Task> taskList = new ArrayList<>();
+        String listLocation = "./data/taskList.txt";
```
Should this file path be saved as a final constant? This would allow you to reference it from a single place, rather than creating multiple of the same variable.

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+
+            } else {
+                addTask(input, taskList, true);
+            }
```
Should the final else statement be used to add a task, or should it be in an "else if" block of its own?

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+                        System.out.println("\nPlease input a valid Date format.\n");
+                    } catch (Exception e) {
+                        System.out.println("\nExcuse me, " +
+                                "please provide a summary and time of event using the tag /at.\n");
```
Should the break be before the '+' operator rather than after?

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+        String reply;
+        try {
+            int index = Integer.parseInt(input.split(" ")[1]);
+            taskList.remove(index-1);
```
Should there be space before and after the '-' operator?

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+        if (indexList.length == 0) {
+            reply += "No index was given . Please provide a valid index.\n";
+            System.out.println(reply);
+        }
```
Should you handle this special condition at the beginning of the method?

hakujitsu commented on src/main/java/Butler.java:
```diff
@@ -0,0 +1,216 @@
+    private static void writeToFile(String textToAdd) throws IOException {
+        String filePath = "./data/taskList.txt";
+        FileWriter fw = new FileWriter(filePath, true);
+        fw.write(textToAdd + System.lineSeparator());
```
Should the user's input be written verbatim into the file?

hakujitsu reviewed (commented):
Your code makes sense! I think you can work on abstracting certain parts out into different methods/classes for greater clarity.

hopinxian committed:
Implement more OOP by extracting Ui, storage, parser, tasklist and commands.

hopinxian committed:
Update ui messages and DukeException messages

hopinxian committed:
Change tasktype from string to enums

hopinxian committed:
Organise classes into Butler packages, which are butler.command, butler.exception, butler.io, butler.task and butler

hopinxian committed:
Add Junit tests

hopinxian committed:
Create jar file

hopinxian committed:
Add JavaDoc comments

hopinxian committed:
Tweak code to comply with coding standard

Code has been modified to follow the Java coding standard at
https://se-education.org/guides/conventions/java/intermediate.html

hopinxian committed:
Add find tasks functionality

Users are able to filter tasks within their task list
by keywords using find function

hopinxian committed:
Merge branch 'branch-A-JavaDoc'

hopinxian committed:
Merge branch 'branch-A-CodingStandard'

Code has been modified to follow the Java coding standard at
https://se-education.org/guides/conventions/java/intermediate.html

hopinxian committed:
Merge branch 'branch-Level-9'

hopinxian committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2021S1/ip into branch-A-Gradle

hopinxian committed:
Update Gradle files and error message

Modifications allow Gradle to run, detect main class
and also allows error messages to be printed.

hopinxian committed:
Remove checkstyle requirement from build.Gradle

hopinxian committed:
Merge branch 'branch-A-Gradle'

hopinxian committed:
Add checkstyle checking

Code has been refactored to follow the Java coding standard.

hopinxian committed:
Merge branch 'branch-A-CheckStyle'

hopinxian committed:
Add basic interactive GUI for Butler

JavaFX is used to set up a GUI with echo functionality.
GUI is added according to Part 1-3 of the JavaFX tutorial.
https://se-education.org/guides/tutorials/javaFx.html

hopinxian committed:
Modify Ui.java to return string responses

Ui.java has been modified to return string responses
instead of printing them out. Scanner is removed from Ui.java.

hopinxian committed:
Add GUI functionality to respond to user input

Command classes now return a String response after execution.
Start method in Butler.java is now able to return the
appropriate response to user input via GUI.

hopinxian committed:
Add FXML to define user interface

hopinxian committed:
Add exit functionality for GUI

hopinxian committed:
Add simple styling

hopinxian committed:
Merge branch 'branch-Level-10'

hopinxian committed:
Update mainClassName within Build.Gradle

hopinxian committed:
Add and enable assertions

hopinxian committed:
Improve code quality

hopinxian committed:
Merge pull request #1 from hopinxian/branch-A-CodeQuality

Improve code quality

hopinxian committed:
Merge branch 'master' into branch-A-Assertions

hopinxian committed:
Merge pull request #2 from hopinxian/branch-A-Assertions

Add and enable assertions

hopinxian committed:
Use streams feature from Java inside TaskList.java

hopinxian committed:
Merge pull request #3 from hopinxian/branch-A-Streams

Use streams feature from Java inside TaskList.java

hopinxian committed:
Add continuous integration

hopinxian committed:
Add undo functionality

hopinxian committed:
Merge branch 'branch-C-Undo'

hopinxian committed:
Add draft user guide

This draft user guide is copied from AddressBook level 3.
Draft upload is for the purpose of viewing user guide on the website.

hopinxian committed:
Update README.md to second revision

hopinxian committed:
Add credit giving statement in README.md

User guide format is referenced from the following link.
https://se-education.org/addressbook-level3/UserGuide.html

hopinxian committed:
Improve GUI to become better

GUI now has circle profile pictures and color background for messages
and scrollpane.

hopinxian committed:
Merge pull request #4 from hopinxian/branch-A-BetterGui

Improve GUI to become better

hopinxian committed:
Set theme jekyll-theme-slate

hopinxian committed:
Update user guide in doc\README.md

Improves formatting and added screenshots.

hopinxian committed:
Merge branch 'master' of https://github.com/hopinxian/ip

hopinxian committed:
Adjust formatting of lists in README.md

hopinxian committed:
Remove the sentence that there is sample data

hopinxian committed:
Add mass ops for delete
